sql/handler.cc

```diff
@@ -6456,6 +6470,91 @@ void handler::set_lock_type(enum thr_lock_type lock)
   table->reginfo.lock_type= lock;
 }
 
+int handler::ha_check_overlaps(const uchar *old_data, const uchar* new_data)
+{
+  DBUG_ASSERT(new_data);
+  if (!table_share->period.unique_keys)
+    return 0;
+
+  bool is_update= old_data != NULL;
+  if (!check_overlaps_buffer)
+    check_overlaps_buffer= (uchar*)alloc_root(&table_share->mem_root,
+                                              table_share->max_unique_length
+                                              + table_share->reclength);
+  auto *record_buffer= check_overlaps_buffer + table_share->max_unique_length;
+  auto *handler= this;
+  if (handler->inited != NONE)
+  {
+    if (!check_overlaps_handler)
+    {
+      check_overlaps_handler= clone(table_share->normalized_path.str,
+                                    &table_share->mem_root);
+      int error= -1;
+      if (check_overlaps_handler != NULL)
+        error= check_overlaps_handler->ha_external_lock(table->in_use, F_RDLCK);
+      if (error)
+        return error;
+    }
+    handler= check_overlaps_handler;
+  }
+
+  for (uint key_nr= 0; key_nr < table_share->keys; key_nr++)
+  {
+    const KEY *key_info= table->key_info + key_nr;
+    if (!key_info->without_overlaps)
+      continue;
+
+    key_copy(check_overlaps_buffer, new_data, key_info, 0);
+    key_part_map keypart_map= (1 << key_info->user_defined_key_parts) - 1;
+    for (auto key_func: {HA_READ_KEY_OR_PREV, HA_READ_KEY_OR_NEXT})
+    {
+      int error= handler->ha_index_read_idx_map(record_buffer, key_nr,
+                                                check_overlaps_buffer,
+                                                keypart_map, key_func);
+      if (error == HA_ERR_KEY_NOT_FOUND)
+        continue;
+      else if (error)
+        return error;
+      else if (is_update
+               && memcmp(old_data + table->s->null_bytes,
```

Comment:
You can avoid index read in case when no keys changed by UPDATE by checking that all key columns are not marked for write. If some key columns are marked for write you can do their field comparison of old and new records (see `compare_record()`).

Reply:
Really, what a messy memcpy, now I see.

I guess I can do just another key_copy, but for old_data, and check equality. That implies less code modifications, no `compare_record` refactoring, or copypasta

Reply:
No, it's ok. I'll add a comment with a description

Reply:
Have been thinking about it more: it is not messy, it is correct. I will add a description

Reply:
It must be avoided anyway. There is `table->write_set` indicating which fields are changed.